### ISSUE TITLE
Incorrect type definition for `Epic`

### ISSUE THREAD

**Issue by Mutefish0** [redux-observable/redux-observable#494]
**Do you want to request a *feature* or report a *bug*?**  
bug

**What is the current behavior?**
```ts
import { Epic } from 'redux-observable'
import { Observable } from 'rxjs'

interface ActionIn {
	type: 'ACTION_IN'
}

interface ActionOut {
	type: 'ACTION_OUT'
	msg: string
}
// Get type error: types of property 'type' are incompatible
const epic: Epic<ActionIn, any, any, ActionOut> = (action$, state) =>  
	Observable.of({
		type: 'ACTION_OUT',
		msg: 'hello'
	} as ActionOut)
```
my solution
```ts
//export declare interface Epic<T extends Action, S, D = any, O extends T = T> {
//  (action$: ActionsObservable<T>, store: MiddlewareAPI<S>, dependencies: D): Observable<O>;
//}
export declare interface Epic<T extends Action, S, D = any, O extends Action = Action> {
  (action$: ActionsObservable<T>, store: MiddlewareAPI<S>, dependencies: D): Observable<O>;
}
```

**What is the expected behavior?**
No type error
**Which versions of redux-observable, and which browser and OS are affected by this issue? Did this work in previous versions of redux-observable?**
"typescript": "^2.8.3"
"redux-observable": "^0.18.0"
"electron": "^2.0.0"

**Comment by DanielRamosAcosta** [comment]
I'm suffering the same type error but with the new version of `redux-observable`

* `typescript`: 2.8.3
* `redux-observable`: 1.0.0-alpha.2
* `rxjs`: `6.2.0`

**Comment by travikk** [comment]
Same issue, cannot have Action<some_type> as input and Action<void> as output, as payloads are incompatible.

**Comment by jayphelps** [comment]
Can someone clarify this issue in the context of 1.0.0-beta.1? https://github.com/redux-observable/redux-observable/blob/69c954b2b2d80891744ed3ada544bd2830553739/index.d.ts#L36-L38

The input must be a superset of the output because in redux-observable any action your epic emits is also recursively provided as the input because epics receive a stream of _all_ actions.

https://stackblitz.com/edit/typescript-vgwtrn
```ts
import { Epic } from 'redux-observable';
import { of } from 'rxjs';

interface ActionIn {
	type: 'ACTION_IN'
}

interface ActionOut {
	type: 'ACTION_OUT'
	msg: string
}

/*
// The input doesn't include the output
const epic: Epic<ActionIn, ActionOut> = (action$, state) =>
  of({
    type: 'ACTION_OUT',
    msg: 'hello'
  } as ActionOut);
*/

// The input includes the output
type Output = ActionOut;
type Input = ActionIn | Output;

const epic: Epic<Input, Output> = (action$, state) =>
  of({
    type: 'ACTION_OUT',
    msg: 'hello'
  } as ActionOut);
```

Since an epic can technically receive all actions, not just those your epic is listening for and the ones it outputs, some people wish to be 100% accurate in their input type by having a union of all actions that can possibly be dispatched somewhere they can import and use in their all of their epics.

**Comment by travikk** [comment]
@jayphelps of course.

To me, the Input Action type should be what is effectively descriminated by ofType. It is especially important for the payload type of action.
So if you have an epic that manages login, you’d have input like: Action<LoginPayload> and output as Action<void | Error> (to notify about success or failure)

Never understood the point of output having the same payload as lower boundary, why would that make sense?
100% of my epics are filtered down by ofType operator at the start, and they produce a completely different action as output.

If people want to produce something that derives from Input Action they should be able to do so, but if they want to produce something totally different that should be allowed too.
In other words, don’t put a constraint “extends input” on output action. That should cater for all our use cases.

**Comment by jayphelps** [comment]
@travikk shoot, I'm not following 🤔can you maybe provide code to illustrate your preferred type definition?

**Comment by travikk** [comment]
@jayphelps sorry, to your point the ideal definition for me, in the context of 1.0.0-beta1 would be:
`export declare interface Epic<T extends Action, O extends Action = Action, S = void, D = any> {
  (action$: ActionsObservable<T>, state$: StateObservable<S>, dependencies: D): Observable<O>;
}`

**Comment by jayphelps** [comment]
Perhaps there's a misunderstanding of what `O extends T` does in this case because of the word "extends"? In this case since we expect `T` and `O` to be unions of `Action`s  it somewhat counterintuitively means the inverse--that `O` is part of the union `T`.

[Playground REPL](http://www.typescriptlang.org/play/#src=declare%20function%20example%3CT%2C%20R%20extends%20T%3E(input%3A%20T)%3A%20R%3B%0D%0A%0D%0Atype%20Output%20%3D%20%7B%20type%3A%20'OUTPUT'%20%7D%3B%0D%0Atype%20Input%20%3D%20%7B%20type%3A%20'INPUT'%20%7D%20%7C%20Output%3B%0D%0A%0D%0Alet%20input%3A%20Input%20%3D%20%7B%20type%3A%20'INPUT'%20%7D%3B%0D%0Alet%20output%3A%20Output%20%3D%20example%3CInput%2C%20Output%3E(input)%3B%0D%0A)

```ts
declare function example<T, R extends T>(input: T): R;

type Output = { type: 'OUTPUT' };
type Input = { type: 'INPUT' } | Output;

let input: Input = { type: 'INPUT' };
let output: Output = example<Input, Output>(input);

// this is fine too and related.
// really think about this and it might
// give you the "a ha" moment :)
input = output;
```

Think of it this way, `<T, R extends T>` is saying that whatever `R` is, I must be able to assign it to a variable of type `T`. 

```ts
function example<T, R extends T>(input: T, output: R): R {
  // this is fine!
  input = output;
  return output;
}
```

```ts
type Output = { type: 'OUTPUT' };
type Input = { type: 'INPUT' } | Output;

let input: Input = { type: 'INPUT' };
let output: Output = { type: 'OUTPUT' };

// also fine
input = output;
```

**Comment by travikk** [comment]
@jayphelps absolutely, the problem starts here however:

```typescript
type Action<TPayload> = {
  type: string;
  payload: T;
};

let input: Action<string> = { type: 'MyInputAction', payload: 'travikk' };
let output: Action<void> = { type: 'OUTPUT' };

// not fine
input = output
```

**Comment by jayphelps** [comment]
@travikk I'm not sure why you would do that though. You lose the `type` field as a tag. i.e. because your `Action<T>` type always has the general `type: string` but the `type` field should be used as the [tag or discriminate](https://en.wikipedia.org/wiki/Tagged_union) making the type uniquely identifiable. The `payload` field may or may not be different--imagine two actions which both have `payload: string`, in that case without the tag you wouldn't be able to distinguish one and the other.
